GitExtensions/StaticDependencyInjection.cs

```diff
@@ -0,0 +1,118 @@
+﻿using GitCommands;
+using System;
+using System.IO;
+using System.Collections.Concurrent;
+
+namespace GitExtensions
+{
+
+    public class DIObjectInstance<T> where T : class, new()
+    {
+        private static object lockObj = new object();
+        private T _Inst;
+        public T Inst
+        {
+            get
+            {
+                if (_Inst == null)
+                {
+                    lock (lockObj)
+                    {
+                        if (_Inst == null)
+                        {
+                            _Inst = StaticDI.CreateInstance<T>(ClearInst);
+                        }
+                    }
+                }
+
+                return _Inst;
+            }
+        }
+
+        private void ClearInst()
+        {
+            lock (lockObj)
+            {
+                _Inst = null;
+            }
+        }
+    }
+
+    public interface IInstanceFactory
+    {
+        T CreateInstance<T>() where T : class, new();
+    }
+
+    public class NewInstanceFactory : IInstanceFactory
+    {
+        public T CreateInstance<T>() where T : class, new()
+        {
+            return new T();
+        }
+    }
+
+    public class StaticDI
```

Comment:
It feels we're reinventing the wheel here. Why not use [Autofac](https://autofac.org/) or alike?

...and if we use DI then interface-based implementation makes even more sense. 
We can do it gradually (no need for a big bang approach) - on per-class level (non-UI). For UI classes (forms and user controls) we can temporarily use service locator pattern (nothing more permanent than temporary things, hey 😆 ). 

Right now we adding a new layer of abstraction where we would have to add a lot of boiler plate code with very little benefit, if any at all. 

Also the API become ugly.

I'm sorry.